File: src/main/java/dev/exit/asset/web/LiabilityController.java
package dev.exit.asset.web;

import java.math.BigDecimal;
import java.util.List;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PositiveOrZero;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import dev.exit.asset.config.CurrentUserProvider;
import dev.exit.asset.domain.Liability;
import dev.exit.asset.domain.LiabilityType;
import dev.exit.asset.portfolio.LiabilityRepository;

@RestController
@RequestMapping("/api/v1/liabilities")
public class LiabilityController {

	private final CurrentUserProvider currentUser;

	private final LiabilityRepository liabilities;

	public LiabilityController(CurrentUserProvider currentUser, LiabilityRepository liabilities) {
		this.currentUser = currentUser;
		this.liabilities = liabilities;
	}

	@GetMapping
	public List<Liability> list() {
		return liabilities.findByOwnerIdOrderByCreatedAtDesc(currentUser.ownerId());
	}

	@PostMapping
	public Liability create(@Valid @RequestBody CreateLiabilityRequest request) {
		return liabilities.save(new Liability(currentUser.ownerId(), request.name(), request.type(),
				request.outstandingBalance(), request.annualInterestRate()));
	}

	public record CreateLiabilityRequest(@NotBlank String name, @NotNull LiabilityType type,
			@NotNull @PositiveOrZero BigDecimal outstandingBalance,
			@NotNull @PositiveOrZero BigDecimal annualInterestRate) {
	}
}

File: liability/src/main/kotlin/dev/exit/asset/liability/LiabilityController.kt
package dev.exit.asset.liability

import dev.exit.asset.common.CurrentUserProvider
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.PositiveOrZero
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.math.BigDecimal

@RestController
@RequestMapping("/api/v1/liabilities")
class LiabilityController(
    private val currentUser: CurrentUserProvider,
    private val liabilities: LiabilityRepository,
) {

    @GetMapping
    fun list(): List<Liability> = liabilities.findByOwnerIdOrderByCreatedAtDesc(currentUser.ownerId())

    @PostMapping
    fun create(@Valid @RequestBody request: CreateLiabilityRequest): Liability =
        liabilities.save(
            Liability(
                currentUser.ownerId(),
                request.name,
                request.type,
                request.outstandingBalance,
                request.annualInterestRate,
            ),
        )

    data class CreateLiabilityRequest(
        @field:NotBlank val name: String,
        @field:NotNull val type: LiabilityType,
        @field:NotNull @field:PositiveOrZero val outstandingBalance: BigDecimal,
        @field:NotNull @field:PositiveOrZero val annualInterestRate: BigDecimal,
    )
}
